src/AIAPI/MainGame/GameCustomFunctionController.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+// using ActionGame;
+using ExtensibleSaveFormat;
+using KKAPI.Maker;
+using UnityEngine;
+using Manager;
+using AIProject;
+using AIChara;
+using CharaCustom;
+
+namespace KKAPI.MainGame
+{
+    /// <summary>
+    /// Base type for custom game extensions.
+    /// It provides many useful methods that abstract away the nasty hooks needed to figure out when the state of the game
+    /// changes.
+    /// 
+    /// This controller is a MonoBehaviour that is created upon registration in <see cref="GameAPI.RegisterExtraBehaviour{T}"/>.
+    /// The controller is created only once. If it's created too late it might miss some events.
+    /// It's recommended to register controllers in your Start method.
+    /// </summary>
+    public abstract class GameCustomFunctionController : MonoBehaviour
+    {
+        private static EnvironmentSimulator _cycle;
+
+        /// <summary>
+        /// Extended save ID used by this function controller
+        /// </summary>
+        public string ExtendedDataId { get; internal set; }
+        
+        /// <summary>
+        /// Get extended data based on supplied ExtendedDataId. When in chara maker loads data from character that's being loaded. 
+        /// </summary>
+        public PluginData GetExtendedData()
+        {
+            if (ExtendedDataId == null) throw new ArgumentException(nameof(ExtendedDataId));
+            return ExtendedSave.GetExtendedDataById(Singleton<CustomBase>.Instance.defChaCtrl, ExtendedDataId);
```

Comment:
Singleton<CustomBase>.Instance.defChaCtrl seems to be the replacement for Manager.Game.Instance.HeroineList in KK.

Reply:
In KK HeroineList is not null anywhere in the main game, it's created at the very start of the app. defChaCtrl might be different, also I remember there were multiple different character lists in different classes, so yeah.
This is what I'm using in CheatTools, and I think this list is the most useful one https://github.com/ManlyMarco/IllusionCheatTools/blob/ae58184eacb0e52155a7a5b0d79de59f99561c3a/AI_CheatTools/CheatToolsWindow.cs#L389

Reply:
Oops this was a replacement for `Manager.Game.Instance.saveData`  So ill just swap it out with  `Singleton<Map>.Instance.Player.ChaControl.chaFile` since I assume it's supposed to be saved to the players card